osu.Game/Database/StandardisedScoreMigrationTools.cs

```diff
@@ -26,6 +26,9 @@ public static bool ShouldMigrateToNewStandardised(ScoreInfo score)
             if (score.IsLegacyScore)
                 return false;
 
+            if (score.TotalScoreVersion > 30000002)
+                return false;
```

Comment:
As already stated, this conditional breaks reimport of pre-score-V2 lazer scores.

https://github.com/ppy/osu/assets/20418176/896730b2-2a8b-4432-9a8a-340737704506

This is because `TotalScoreVersion` can _only_ be `30000002` if the score was migrated from an older release via realm migrations. Otherwise it will _always_ be `30000003` due to

https://github.com/ppy/osu/blob/1868826d692f9d0c0fc13a9184758a1c3518a85f/osu.Game/Scoring/ScoreInfo.cs#L64-L74

Two possible fixes I can see:

## Option 1 ("the conservative", but also "the confusing")

```diff
diff --git a/osu.Game/Scoring/Legacy/LegacyScoreDecoder.cs b/osu.Game/Scoring/Legacy/LegacyScoreDecoder.cs
index c6461840aa..96188ed24f 100644
--- a/osu.Game/Scoring/Legacy/LegacyScoreDecoder.cs
+++ b/osu.Game/Scoring/Legacy/LegacyScoreDecoder.cs
@@ -48,6 +48,7 @@ public Score Parse(Stream stream)
                 int version = sr.ReadInt32();
 
                 scoreInfo.IsLegacyScore = version < LegacyScoreEncoder.FIRST_LAZER_VERSION;
+                scoreInfo.TotalScoreVersion = version < 30000002 ? 30000001 : LegacyScoreEncoder.LATEST_VERSION;
 
                 string beatmapHash = sr.ReadString();
 

```

This works by checking whether the replay was set prior to https://github.com/ppy/osu/pull/23848, and setting an arbitrary score version lower than 30000002 (see how [having two disparate versions that mean different things, but share the same numerical schemes, gets confusing?](https://github.com/ppy/osu/pull/24072#discussion_r1245668884))

## Option 2 ("the nuclear")

```diff
diff --git a/osu.Game/Database/StandardisedScoreMigrationTools.cs b/osu.Game/Database/StandardisedScoreMigrationTools.cs
index 11a5e252a3..f2a7f5dfe0 100644
--- a/osu.Game/Database/StandardisedScoreMigrationTools.cs
+++ b/osu.Game/Database/StandardisedScoreMigrationTools.cs
@@ -26,15 +26,7 @@ public static bool ShouldMigrateToNewStandardised(ScoreInfo score)
             if (score.IsLegacyScore)
                 return false;
 
-            if (score.TotalScoreVersion > 30000002)
-                return false;
-
-            // Recalculate the old-style standardised score to see if this was an old lazer score.
-            bool oldScoreMatchesExpectations = GetOldStandardised(score) == score.TotalScore;
-            // Some older scores don't have correct statistics populated, so let's give them benefit of doubt.
-            bool scoreIsVeryOld = score.Date < new DateTime(2023, 1, 1, 0, 0, 0);
-
-            return oldScoreMatchesExpectations || scoreIsVeryOld;
+            return score.Date < new DateTime(2023, 6, 5, 0, 0, 0);
         }
 
         public static long GetNewStandardised(ScoreInfo score)
@@ -166,35 +158,6 @@ private static HitResult getMaxJudgementFor(HitResult hitResult, HitResult max)
             return HitResult.IgnoreHit;
         }
 
-        public static long GetOldStandardised(ScoreInfo score)
-        {
-            double accuracyScore =
-                (double)score.Statistics.Where(kvp => kvp.Key.AffectsAccuracy()).Sum(kvp => Judgement.ToNumericResult(kvp.Key) * kvp.Value)
-                / score.MaximumStatistics.Where(kvp => kvp.Key.AffectsAccuracy()).Sum(kvp => Judgement.ToNumericResult(kvp.Key) * kvp.Value);
-            double comboScore = (double)score.MaxCombo / score.MaximumStatistics.Where(kvp => kvp.Key.AffectsCombo()).Sum(kvp => kvp.Value);
-            double bonusScore = score.Statistics.Where(kvp => kvp.Key.IsBonus()).Sum(kvp => Judgement.ToNumericResult(kvp.Key) * kvp.Value);
-
-            double accuracyPortion = 0.3;
-
-            switch (score.RulesetID)
-            {
-                case 1:
-                    accuracyPortion = 0.75;
-                    break;
-
-                case 3:
-                    accuracyPortion = 0.99;
-                    break;
-            }
-
-            double modMultiplier = 1;
-
-            foreach (var mod in score.Mods)
-                modMultiplier *= mod.ScoreMultiplier;
-
-            return (long)Math.Round((1000000 * (accuracyPortion * accuracyScore + (1 - accuracyPortion) * comboScore) + bonusScore) * modMultiplier);
-        }
-
         /// <summary>
         /// Converts from <see cref="ScoreInfo.LegacyTotalScore"/> to the new standardised scoring of <see cref="ScoreProcessor"/>.
         /// </summary>

```

This works by removing the entire "guess if this is an old score by recalculating it using the old algorithm" machinery that is causing this to be reached via gameplay flows, and instead just basing it on the hard cutoff date of introducing Score V2.

---

Would be interested to hear @smoogipoo's take on this. I will say I'm biased towards the nuclear option here.

Reply:
The problem with the second option is that for a user that may not have updated to the latest version when it was released, all their recent scores will be misidentified.

Reply:
I've gone with option 1 due to above reasoning, and added inline commentary. Please check it makes sense.

Reply:
Looks fine to me :+1: